src/backend/api/meals.js

```diff
@@ -1,12 +1,166 @@
+const { query } = require("express");
 const express = require("express");
+const { where } = require("../database");
 const router = express.Router();
 const knex = require("../database");
 
 router.get("/", async (request, response) => {
   try {
-    // knex syntax for selecting things. Look up the documentation for knex for further info
-    const titles = await knex("meals").select("title");
-    response.json(titles);
+    let mealArrayToChange = await knex
+      .raw(
+        `SELECT meal.id,meal.title, meal.description, meal.location,meal.when_time, meal.price, meal.created_date,
+      meal.max_reservations,
+      COALESCE(SUM(reservation.number_of_guests),0) AS total_reserved
+      FROM   meal
+      LEFT JOIN reservation
+      ON meal.id = reservation.meal_id
+      GROUP  BY meal.id`
+      )
+      .then((result) => result[0]);
+
+    if (Object.keys(request.query).length === 0) {
+      response.send(mealArrayToChange);
+      return;
+    }
+
+    if (typeof request.query["availableReservations"] !== "undefined") {
+      if (request.query["availableReservations"] === "true") {
+        mealArrayToChange = mealArrayToChange.filter(
+          (item) => item.total_reserved < item.max_reservations
+        );
+      } else if (request.query["availableReservations"] === "false") {
+        mealArrayToChange = mealArrayToChange.filter(
+          (item) =>
+            Number(item.total_reserved) === Number(item.max_reservations)
+        );
+      } else {
+        response
+          .status(400)
+          .json({ error: "Invalid availableReservations param" });
+        return;
+      }
+    }
+
+    if (request.query["maxPrice"]) {
+      if (!isNaN(Number(request.query["maxPrice"]))) {
+        mealArrayToChange = mealArrayToChange.filter(
+          (item) => item.price < Number(request.query["maxPrice"])
+        );
+      } else {
+        response.status(400).json({ error: "Invalid maxPrice param" });
+        return;
+      }
+    }
+
+    if (request.query["title"]) {
+      mealArrayToChange = mealArrayToChange.filter((item) =>
+        item.title.toLowerCase().includes(request.query["title"].toLowerCase())
+      );
+    }
+
+    if (request.query["createdAfter"]) {
+      const createdAfter = new Date(request.query["createdAfter"]);
+      if (createdAfter != "Invalid Date") {
+        mealArrayToChange = mealArrayToChange.filter((item) => {
+          const itemDate = new Date(item.createdAt);
+          return itemDate > createdAfter;
+        });
+      } else {
```

Comment:
Please consider using knex chainable methods for selection and filtering.

Reply:
@jslambda
This way to make query not working in some reason. "Having" part does not act as expected.

  queryToMeal = queryToMeal
    .sum({ totalReserved: "reservation.number_of_guests" })
    .join("reservation", "reservation.meal_id", "=", "meal.id")
    .groupBy("meal.id")
    .having('totalReserved', '<','meal.max_reservations'); // 0 objects return here